common/rfb/H264LibavDecoderContext.cxx

```diff
@@ -222,9 +222,19 @@ void H264LibavDecoderContext::decode(const uint8_t* h264_in_buffer,
 
   int stride;
   pb->getBuffer(rect, &stride);
-  int dst_linesize = stride * pb->getPF().bpp/8;  // stride is in pixels, linesize is in bytes (stride x4). We need bytes
 
-  sws_scale(sws, frame->data, frame->linesize, 0, frame->height, &swsBuffer, &dst_linesize);
+  int height = frame->height;
 
-  pb->imageRect(rect, swsBuffer, stride);
+  // The server may send a frame with a height greater than that of the rect,
+  // but we don't want to write outside of the rect.
+  // The server may choosed to do this due to hardware encoder constraints.
+  if (height > rect.height())
+    height = rect.height();
```

Comment:
Why not width as well?

Reply:
Because this fixed the crashing and adding the width too was a bit more work and it was time to go to bed.

Reply:
Actually, I think that this might be handled by the linesize arguments to `sws_scale`, but I'd have to take a closer look at the implementation of `sws_scale` to be sure.

Reply:
Please check. It would be very odd to specify that height can be larger, but not width.

Reply:
Well, `sws_scale` is quite a complex beast. It can go into multiple code paths depending on the pixel formats and there are different implementations for different platforms.

It's probably safest to go by the guarantees given by the API, which are none.

A better approach would be to make the scaling buffer as wide as the frame and handle the cropping in `pb->imageRect` instead. E.g. by setting the stride.

Reply:
Either way is fine by me.

Reply:
Current iteration takes care of width, height and alignment.

Reply:
Well I do not think the current proposed version resolves the swsBuffer size issue that I see. 

sws_scale wants that there is room after end of the buffer(I found this by experimentation) , which we need to ignore .. I assume as you said related to SIMD code that is optimized for CPU/platforms and it needs to write in chunks of data , so it needs to have room at end for that garbage data.

new proposed code:

  int dstLineSize = FFALIGN(frame->width * dstBytesPerPixel , av_cpu_max_align());                                                                             
  swsBuffer_test.resize(dstLineSize * frame->height, 0);                                                                                                       
  printf("swsBuffer_test.capacity=%d\n",swsBuffer_test.capacity());                                                                                            

result: swsBuffer_test.capacity=4202496 
4202496 is exactly 1368*4*768 (1368x768 is my frame)

When I allocate the buffer with av_frame_get_buffer on a AVFrame , the buffer have extra room (visible with mtrace/muntrace, I was not able to find a better way to figure size of allocated buffer ):

         mtrace();                                                                                                                                             
         av_frame_get_buffer( pic_out, av_cpu_max_align() );                                         
         muntrace();                                                                                                                                           

LD_PRELOAD=/usr/lib/libc_malloc_debug.so MALLOC_TRACE=/tmp/x ./test
cat /tmp/x
= Start
@ /usr/lib/libavutil.so.59:(av_malloc+66)[0x499c6] + 0x741a595c8040 0x402080

0x402080 == 1368*4*768 + 128  . 
The issue that I am facing is .. sws_scale writes garbage data after end of the buffer so we  need some extra room for that. Not sure how much space we need (I guess it depends on CPU/etc) ,we could create a frame  with av_frame_alloc  / av_frame_get_buffer and use that as the dest buffer for sws_scale then free it ? Or just allocate that once and keep it untill the frame size changes?

Reply:
Yeah, using `AVFrame` for the scaling buffer is probably a good idea as it handles the alignment for us.

The calculation of the frame size a slightly more complex than what we have, but it adds something called `plane_padding`, which is `(16 + 16) * 4 = 128`.

See: https://ffmpeg.org/doxygen/trunk/frame_8c_source.html#l00169

There is one problem with the current approach (and using `AVFrame`). Technically, after the stride is aligned, it might not be divisible by bytes-per-pixel. The solution for that would be to either align on the width or to add a new overload for `imageRect` that takes stride in bytes rather than row pixel count.

Reply:
At a glance it looks good to me. Will test it a bit on my setup ..

Also to get rid of the memleak of sws from sws_getContext  , need to freeContext(sws) after sws_scale* (valgrind shows it)

Reply:
So far testing looks good. doing more testing

Reply:
OK so this sws_freeContext(sws) needs to be done in H264LibavDecoderContext::freeCodec()

Reply:
So far no crash in my testing. However there is some big memleak(when h264 resets),trying to figure why

Reply:
I added a commit to clean up `sws`.

Reply:
One more thing, for sws_getCachedContext , we passing flags 0 , however in ffmpeg code that will be converted to SWS_BICUBIC , which in turn means every time a new SwsContext will be allocated unless we pass flags SWS_BICUBIC like this..

  sws = sws_getCachedContext(sws, frame->width, frame->height, avctx->pix_fmt,                                                                                 
                             frame->width, frame->height, AV_PIX_FMT_RGB32,                                                                                    
                             SWS_BICUBIC, nullptr, nullptr, nullptr);

Reply:
Well, I'd say that it shouldn't be using any interpolation at all; just use nearest neighbour, but it seems that this is not available as an option, so I suppose `SWS_FAST_BILINEAR` would be the least bad option.

In any case, this is a separate issue.

Reply:
Right it is a separate issue,however right now we are using sws_getCachedContext and it does not work as expected, is always going to return a new context and free previous one because 0 is translated to 4 (SWS_BICUBIC) inside sws_init_single_context . 

I understand this needs to be fixed in a diff PR.

BTW So far all my testing shows good results. I see there is an issue with the build(sws_scale_frame was introduced in 2021 Sep based on ffmpeg doc/APIchanges)